components/ContainedInputViews/src/MDCContainedInputView.h

```diff
@@ -0,0 +1,319 @@
+ */
+@property(strong, nonatomic, readonly, nonnull) UILabel *leadingUnderlineLabel;
+
+/**
+ The @c trailingUnderlineLabel can be used to display helper or error text.
+ */
+@property(strong, nonatomic, readonly, nonnull) UILabel *trailingUnderlineLabel;
+
+/**
+ This property is used to determine how much horizontal space to allot for each of the two underline
+ labels.
+
+ @note The default value is MDCContainedInputViewUnderlineLabelDrawPriorityTrailing. The rationale
+ behind this is it is less likely to have long explanatory error text and more likely to have short
+ text, like a character counter. It is better to draw the short text first and use whatever space is
+ leftover for the longer text, which may wrap to new lines.
+ */
+@property(nonatomic, assign)
+    MDCContainedInputViewUnderlineLabelDrawPriority underlineLabelDrawPriority;
+
+/**
+ When @c underlineLabelDrawPriority is set to @c .custom the value of this property helps determine
+ what percentage of the available width each underline label gets. It can be thought of as a
+ divider. A value of @c 0 would result in the trailing underline label getting all the available
+ width. A value of @c 1 would result in the leading underline label getting all the available width.
+ A value of @c .5 would result in each underline label getting 50% of the available width.
+ */
+@property(nonatomic, assign) CGFloat customUnderlineLabelDrawPriority;
+
+/**
+ When set to YES, the floating label floats when the view becomes the first responder. When
+ set to NO it disappears.
+
+ @note The default is YES.
+ */
+@property(nonatomic, assign) BOOL canFloatingLabelFloat;
+
+/**
+ This property toggles the error state (similar to @c isHighlighted, @c isEnabled, @c isSelected,
+ etc.) that is part of a general interpretation of the states outlined in the Material guidelines
+ for Text Fields. See the @c MDCContainedInputViewState enum for more information.
+ */
+@property(nonatomic, assign) BOOL isErrored;
+
+/**
+ This property toggles the activated state (similar to @c isHighlighted, @c isEnabled, @c
+ isSelected, etc.) that is part of a general interpretation of the states outlined in the Material
+ guidelines for Text Fields. See the @c MDCContainedInputViewState enum for more information.
+ */
+@property(nonatomic, assign) BOOL isActivated;
+
+/**
+  This method returns a color scheme for a given state.
+ */
+- (nonnull id<MDCContainedInputViewColorScheming>)containedInputViewColorSchemingForState:
+    (MDCContainedInputViewState)containedInputViewState;
+
+/**
+ This method sets a color scheme for a given state.
+ */
+- (void)setContainedInputViewColorScheming:
+            (nonnull id<MDCContainedInputViewColorScheming>)containedInputViewColorScheming
+                                  forState:(MDCContainedInputViewState)textFieldState;
+
+/**
+ Returns the rect surrounding the main content, i.e. the area that the container should be drawn
+ around.
```

Comment:
Tip: Explain how clients should use this. What is this for?

Reply:
This is actually only here because the styles need to know where to draw the containers. I couldn't think of a better way to give the container style objects this information.